File: scripts/download_new_images.py
#!/usr/bin/env python3
import requests
import os
from PIL import Image
import io

# Create images directory if it doesn't exist
os.makedirs('images', exist_ok=True)

# Subject images with Wikimedia Commons URLs optimized for lesson cards
subjects = {
    'mathematics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Mathematics_formula.svg',
        'description': 'Mathematics symbols and formulas',
        'license': 'Public Domain'
    },
    'english-literature.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Books-aj.svg_aj_ashton_01.svg',
        'description': 'Stack of books representing literature',
        'license': 'CC BY-SA 3.0'
    },
    'chemistry.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Chemistry_lab.svg',
        'description': 'Chemistry laboratory equipment',
        'license': 'Public Domain'
    },
    'physics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Physics.svg',
        'description': 'Physics symbols and equations',
        'license': 'Public Domain'
    },
    'computer-science.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Computer_code.svg',
        'description': 'Computer code and programming symbols',
        'license': 'Public Domain'
    },
    'biology.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Dna.svg',
        'description': 'DNA double helix structure',
        'license': 'Public Domain'
    },
    'art-design.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Palette.svg',
        'description': 'Artist palette with colors',
        'license': 'Public Domain'
    },
    'music-theory.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Musical_notes.svg',
        'description': 'Musical notes and staff',
        'license': 'Public Domain'
    },
    'economics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Finance_chart.svg',
        'description': 'Economic chart and graph',
        'license': 'Public Domain'
    },
    'psychology.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Brain.svg',
        'description': 'Human brain diagram',
        'license': 'Public Domain'
    }
}

def download_and_resize_image(filename, data):
    try:
        # Extract the actual filename from the URL
        file_name = data['url'].split('/')[-1].replace(':', '_')
        
        # Use the correct Wikimedia Commons download URL format
        download_url = f"https://upload.wikimedia.org/wikipedia/commons/{file_name}"
        
        print(f"Downloading {filename} from {download_url}")
        
        # Add headers to mimic a browser request
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        response = requests.get(download_url, timeout=30, headers=headers)
        response.raise_for_status()
        
        # Open image and resize for lesson cards (optimal around 300x200)
        img = Image.open(io.BytesIO(response.content))
        
        # Convert to RGB if necessary (for PNG with transparency)
        if img.mode in ('RGBA', 'LA', 'P'):
            background = Image.new('RGB', img.size, (255, 255, 255))
            if img.mode == 'P':
                img = img.convert('RGBA')
            background.paste(img, mask=img.split()[-1] if img.mode == 'RGBA' else None)
            img = background
        
        # Resize to maintain aspect ratio while fitting lesson card dimensions
        max_width, max_height = 400, 300
        img.thumbnail((max_width, max_height), Image.Resampling.LANCZOS)
        
        # Save as JPEG with good quality
        img_path = f'images/{filename}'
        img.save(img_path, 'JPEG', quality=85, optimize=True)
        
        # Get file size
        file_size = os.path.getsize(img_path) / 1024  # KB
        
        print(f"✓ Downloaded and resized {filename} ({file_size:.1f} KB)")
        
        return {
            'source': data['url'],
            'license': data['license'],
            'description': data['description'],
            'size_kb': round(file_size, 1)
        }
        
    except Exception as e:
        print(f"✗ Failed to download {filename}: {e}")
        return None

# Download and process all images
sources_data = {}

for filename, data in subjects.items():
    result = download_and_resize_image(filename, data)
    if result:
        sources_data[filename] = result

# Save sources.json
import json
with open('images/sources.json', 'w') as f:
    json.dump(sources_data, f, indent=2)

print(f"\n✓ Downloaded {len(sources_data)} images successfully")
print("✓ Sources saved to images/sources.json")

File: scripts/download_new_images.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const axios = require('axios');
const sharp = require('sharp');

// Create images directory if it doesn't exist
const imagesDir = 'images';
if (!fs.existsSync(imagesDir)) {
    fs.mkdirSync(imagesDir, { recursive: true });
}

// Subject images with Wikimedia Commons URLs optimized for lesson cards
const subjects = {
    'mathematics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Mathematics_formula.svg',
        'description': 'Mathematics symbols and formulas',
        'license': 'Public Domain'
    },
    'english-literature.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Books-aj.svg_aj_ashton_01.svg',
        'description': 'Stack of books representing literature',
        'license': 'CC BY-SA 3.0'
    },
    'chemistry.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Chemistry_lab.svg',
        'description': 'Chemistry laboratory equipment',
        'license': 'Public Domain'
    },
    'physics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Physics.svg',
        'description': 'Physics symbols and equations',
        'license': 'Public Domain'
    },
    'computer-science.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Computer_code.svg',
        'description': 'Computer code and programming symbols',
        'license': 'Public Domain'
    },
    'biology.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Dna.svg',
        'description': 'DNA double helix structure',
        'license': 'Public Domain'
    },
    'art-design.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Palette.svg',
        'description': 'Artist palette with colors',
        'license': 'Public Domain'
    },
    'music-theory.png': {
        'url': 'https://commons.wikimedia.org/wiki/File:Musical_notes.svg',
        'description': 'Musical notes and staff',
        'license': 'Public Domain'
    },
    'economics.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Finance_chart.svg',
        'description': 'Economic chart and graph',
        'license': 'Public Domain'
    },
    'psychology.jpg': {
        'url': 'https://commons.wikimedia.org/wiki/File:Brain.svg',
        'description': 'Human brain diagram',
        'license': 'Public Domain'
    }
};

async function downloadAndResizeImage(filename, data) {
    try {
        // Extract the actual filename from the URL
        const fileName = data.url.split('/').pop().replace(/:/g, '_');
        
        // Use the correct Wikimedia Commons download URL format
        const downloadUrl = `https://upload.wikimedia.org/wikipedia/commons/${fileName}`;
        
        console.log(`Downloading ${filename} from ${downloadUrl}`);
        
        // Add headers to mimic a browser request
        const headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        };
        
        const response = await axios({
            method: 'get',
            url: downloadUrl,
            responseType: 'arraybuffer',
            timeout: 30000,
            headers: headers
        });
        
        // Process image with sharp
        const imgBuffer = Buffer.from(response.data);
        
        // Resize to maintain aspect ratio while fitting lesson card dimensions
        const maxWidth = 400;
        const maxHeight = 300;
        
        const outputPath = path.join(imagesDir, filename);
        await sharp(imgBuffer)
            .resize(maxWidth, maxHeight, { 
                fit: 'inside',
                withoutEnlargement: true 
            })
            .jpeg({ quality: 85, progressive: true })
            .toFile(outputPath);
        
        // Get file size
        const stats = fs.statSync(outputPath);
        const fileSizeKB = stats.size / 1024;
        
        console.log(`✓ Downloaded and resized ${filename} (${fileSizeKB.toFixed(1)} KB)`);
        
        return {
            source: data.url,
            license: data.license,
            description: data.description,
            size_kb: Math.round(fileSizeKB * 10) / 10
        };
        
    } catch (error) {
        console.log(`✗ Failed to download ${filename}: ${error.message}`);
        return null;
    }
}

async function downloadAllImages() {
    const sourcesData = {};
    
    for (const [filename, data] of Object.entries(subjects)) {
        const result = await downloadAndResizeImage(filename, data);
        if (result) {
            sourcesData[filename] = result;
        }
    }
    
    // Save sources.json
    const sourcesPath = path.join(imagesDir, 'sources.json');
    fs.writeFileSync(sourcesPath, JSON.stringify(sourcesData, null, 2));
    
    console.log(`\n✓ Downloaded ${Object.keys(sourcesData).length} images successfully`);
    console.log('✓ Sources saved to images/sources.json');
}

downloadAllImages().catch(console.error);
